pkg/segment/writer/segmetarw.go

```diff
@@ -480,12 +494,98 @@ func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
 	if sfmData.AllPQIDs == nil {
 		sfmData.AllPQIDs = make(map[string]bool)
 	}
-	sfmData.AllPQIDs[newpqid] = true
 
-	// TODO this should be through a channel so that we only have one writer
+	utils.MergeMapsRetainingFirst(sfmData.AllPQIDs, pqidMap)
+
 	writeSfm(segkey, sfmData)
 }
 
+func BackFillPQSSegmetaEntry(segkey string, newpqid string) {
+	BulkBackFillPQSSegmetaEntries(segkey, map[string]bool{newpqid: true})
+}
+
+// AddToBackFillAndEmptyPQSChan adds a new pqid to the channel
+// if emptyPqs is true, then it will also add the EmptyResults for this pqid
+func AddToBackFillAndEmptyPQSChan(segkey string, newpqid string, emptyPqs bool) {
+	pqsChan <- PQSChanMeta{pqid: newpqid, segKey: segkey, emptyPqs: emptyPqs}
+}
+
+func listenBackFillAndEmptyPQSRequests() {
+	// Listen on the channel, every PQS_TICKER seconds or if the size of the channel is PQS_FLUSH_SIZE,
+	// it would get all the data in the channel and then do the process of Backfilling PQMR files.
+	// This is to avoid multiple writes to the same file
+
+	ticker := time.NewTicker(PQS_TICKER * time.Second) // every 10 seconds
+	defer ticker.Stop()
+
+	buffer := make([]PQSChanMeta, 0, PQS_FLUSH_SIZE)
+
+	callProcessBackFillPQSRequests := func() {
+		processBackFillAndEmptyPQSRequests(buffer)
+		buffer = buffer[:0]
+	}
+
+	for {
+		select {
+		case pqsChanMeta := <-pqsChan:
+			buffer = append(buffer, pqsChanMeta)
+			if len(buffer) >= PQS_FLUSH_SIZE {
+				callProcessBackFillPQSRequests()
+			}
+		case <-ticker.C:
+			if len(buffer) > 0 {
+				callProcessBackFillPQSRequests()
+			}
+		}
+	}
+}
+
+func processBackFillAndEmptyPQSRequests(pqsRequests []PQSChanMeta) {
+	if len(pqsRequests) == 0 {
+		return
+	}
+
+	// segKey -> pqid -> true
+	segKeyPqidMap := make(map[string]map[string]bool)
+	// pqid -> segKey -> true
+	pqidSegKeyMap := make(map[string]map[string]bool) // for empty PQS
+
+	for _, pqsRequest := range pqsRequests {
+		if _, ok := segKeyPqidMap[pqsRequest.segKey]; !ok {
+			segKeyPqidMap[pqsRequest.segKey] = make(map[string]bool)
+		}
+		segKeyPqidMap[pqsRequest.segKey][pqsRequest.pqid] = true
```

Comment:
shouldn't we be adding to the `segKeyPqidMap` only if `emptyPqs` is false ? 
also I prefer the maps to be called as what they mean, for e.g. one map means you are writing `this segment has this valid pqs data` while the other means `this pqs id contains these segments which have no-matching results or emptyPQS data`.

Reply:
No We would add to `segKeyPqidMap` irrespective of whether `emptyPQS` is set to `true` or `false`. As this `segKeytoPQID` map is the one which will be updated per segment in the `sfm` file. So that one will add all the `Non-empty` and `Empty` PQIDs.

So the first map contains all the PQIDs for that segment (empty or non-empty). The second map is a map from PQID to seg key which represents the empty segments for a PQID.

I will try to add some better names.